lib/chai/core/assertions.js

```diff
@@ -1268,120 +1268,114 @@ module.exports = function (chai, _) {
    */
 
   function assertThrows (constructor, errMsg, msg) {
-    if (msg) flag(this, 'message', msg);
-    var obj = flag(this, 'object');
-    new Assertion(obj, msg).is.a('function');
 
-    var thrown = false
-      , desiredError = null
-      , name = null
-      , thrownError = null;
-
-    if (arguments.length === 0) {
-      errMsg = null;
-      constructor = null;
-    } else if (constructor && (constructor instanceof RegExp || 'string' === typeof constructor)) {
-      errMsg = constructor;
-      constructor = null;
-    } else if (constructor && constructor instanceof Error) {
-      desiredError = constructor;
-      constructor = null;
-      errMsg = null;
-    } else if (typeof constructor === 'function') {
-      name = constructor.prototype.name;
-      if (!name || (name === 'Error' && constructor !== Error)) {
-        name = constructor.name || (new constructor()).name;
+    function stringify (expected) {
+      if (expected.e) {
+        return expected.e.message ? expected.e.toString() : expected.e.name;
+      }
```

Comment:
Hey @jails could you please explain why we check `expected.e` - it's not a convention I'm familiar with.